core/include/vecmem/utils/impl/copy.ipp

```diff
@@ -395,6 +398,138 @@ copy::event_type copy::set_sizes(
     return create_event();
 }
 
+template <typename SCHEMA>
+copy::event_type copy::setup(edm::view<SCHEMA> data) const {
+
+    // For empty containers nothing needs to be done.
+    if (data.capacity() == 0) {
+        return vecmem::copy::create_event();
+    }
+
+    // Copy the data layout to the device, if needed.
+    if (data.layout().ptr() != data.host_layout().ptr()) {
+        operator()(data.host_layout(), data.layout(), type::unknown);
+    }
+
+    // Initialize the "size variable(s)" correctly on the buffer.
+    if (data.size().ptr() != nullptr) {
+        memset(data.size(), 0);
+    }
+    VECMEM_DEBUG_MSG(3,
+                     "Prepared an SoA container of capacity %u "
+                     "for use on a device (layout: {%u, %p}, size: {%u, %p})",
+                     data.capacity(), data.layout().size(),
+                     static_cast<void*>(data.layout().ptr()),
+                     data.size().size(), static_cast<void*>(data.size().ptr()));
+
+    // Return a new event.
+    return create_event();
+}
+
+template <typename... VARTYPES>
+copy::event_type copy::memset(edm::view<edm::schema<VARTYPES...>> data,
+                              int value) const {
+
+    // For buffers, we can do this in one go.
+    if (data.payload().ptr() != nullptr) {
+        memset(data.payload(), value);
+    } else {
+        // Do the operation using the helper function, recursively.
+        memset_impl(data, value, std::index_sequence_for<VARTYPES...>{});
+    }
+
+    // Return a new event.
+    return create_event();
+}
+
+template <typename... VARTYPES1, typename... VARTYPES2>
+copy::event_type copy::operator()(
+    const edm::view<edm::schema<VARTYPES1...>>& from_view,
+    edm::view<edm::schema<VARTYPES2...>> to_view,
+    type::copy_type cptype) const {
+
+    // The input and output types are allowed to be different, but only by
+    // const-ness.
+    static_assert(sizeof...(VARTYPES1) == sizeof...(VARTYPES2),
+                  "Can only use compatible types in the copy");
+    static_assert(
+        std::conjunction_v<std::is_same<VARTYPES1, VARTYPES2>...> ||
+            std::conjunction_v<
+                edm::type::details::is_same_nc<VARTYPES1, VARTYPES2>...>,
+        "Can only use compatible types in the copy");
+
+    // First, handle the simple case, when both views have a contiguous memory
+    // layout.
+    if ((from_view.payload().ptr() != nullptr) &&
```

Comment:
This was a little bit hinted at in #254, but I didn't write it down really explicitly.

Let's start with how we currently determine contiguity for jagged vectors. That is done here:

https://github.com/acts-project/vecmem/blob/main/core/include/vecmem/utils/impl/copy.ipp#L528-L544

While it may look correct, it's actually kind of fragile. See:

https://forums.developer.nvidia.com/t/single-cudamemcpy-across-multiple-allocations

Under the right (or wrong...) circumstances that logic can mis-categorize a jagged vector view. :frowning: (It's just circumstantial enough that I still didn't fix it.)

At first I thought that I would write some similar, albeit more complicated logic for checking if an entire SoA container is contiguous in memory. Based on the pointers of the views of the individual variables in the container. But while for jagged vectors this logic is already fragile enough, for SoA containers it would be super brittle. Because: alignment!

In `vecmem::edm::buffer` I ended up using [vecmem::details::aligned_multiple_placement](https://github.com/acts-project/vecmem/blob/main/core/include/vecmem/containers/details/aligned_multiple_placement.hpp). This is because when allocating space for a bunch of arrays of different types, the very end of one array's block may not be the correct starting position for the next array's block. So even if a container is held by `vecmem::edm::buffer` in a single memory block, the individual views inside the `vecmem::edm::view` object may not be pointing at contiguous spaces in memory.

I considered a number of horrible solutions to this, but in the end decided to rather just add some variables to `vecmem::edm::view` that are **only** set by `vecmem::edm::buffer`. The `payload()` view is one of these. When constructing a `vecmem::edm::view` from a `vecmem::edm::host` container, `vecmem::edm::view::payload()` will never point to anything. Only when `vecmem::edm::view` describes a buffer, does `vecmem::edm::view::payload()` point to a non-null memory address. And since the memory allocated by `vecmem::edm::buffer` is always contiguous by construction, if `payload()` is "valid", the copy code can know that it can just copy things as they are. :wink:

I could continue the description a little longer as well, but hopefully this conveys the main idea. :wink: